kvdb/postgres/readwrite_tx.go

```diff
@@ -107,20 +128,83 @@ func (tx *readWriteTx) Rollback() error {
 // ReadWriteBucket opens the root bucket for read/write access.  If the
 // bucket described by the key does not exist, nil is returned.
 func (tx *readWriteTx) ReadWriteBucket(key []byte) walletdb.ReadWriteBucket {
-	return rootBucket(tx).NestedReadWriteBucket(key)
+	table := toTableName(key)
+
+	var value int
+	err := tx.tx.QueryRow(
+		context.TODO(),
+		"SELECT 1 FROM pg_catalog.pg_tables WHERE schemaname='public' and tablename=$1",
+		table,
+	).Scan(&value)
+
+	if err == pgx.ErrNoRows {
+		return nil
+	}
+
+	return newReadWriteBucket(tx, table, nil)
 }
 
 // CreateTopLevelBucket creates the top level bucket for a key if it
 // does not exist.  The newly-created bucket it returned.
 func (tx *readWriteTx) CreateTopLevelBucket(key []byte) (walletdb.ReadWriteBucket, error) {
-	return rootBucket(tx).CreateBucketIfNotExists(key)
+	table := toTableName(key)
+
+	_, err := tx.tx.Exec(context.TODO(), `
+	
+CREATE TABLE IF NOT EXISTS public.`+table+`
+(
+    key bytea NOT NULL,
+    value bytea,
+    parent_id bigint,
+    id bigserial PRIMARY KEY,
+    sequence bigint,
+    CONSTRAINT `+table+`_parent FOREIGN KEY (parent_id)
+        REFERENCES public.`+table+` (id) MATCH SIMPLE
+        ON UPDATE NO ACTION
+        ON DELETE CASCADE
+        NOT VALID
+);
+
+CREATE INDEX IF NOT EXISTS `+table+`_p
+    ON public.`+table+` USING btree
+    (parent_id ASC NULLS LAST);
+
+CREATE UNIQUE INDEX IF NOT EXISTS `+table+`_up
+    ON public.`+table+` USING btree
+    (parent_id ASC NULLS LAST, key ASC NULLS LAST) WHERE parent_id IS NOT NULL;
+
+CREATE UNIQUE INDEX IF NOT EXISTS `+table+`_unp ON public.`+table+` (key) WHERE parent_id IS NULL;
+
+`)
+	if err != nil {
+		return nil, err
+	}
+
+	return newReadWriteBucket(tx, table, nil), nil
 }
 
 // DeleteTopLevelBucket deletes the top level bucket for a key.  This
 // errors if the bucket can not be found or the key keys a single value
 // instead of a bucket.
 func (tx *readWriteTx) DeleteTopLevelBucket(key []byte) error {
-	return rootBucket(tx).DeleteNestedBucket(key)
+	bucket := tx.ReadWriteBucket(key)
+	if bucket == nil {
+		return walletdb.ErrBucketNotFound
+	}
+
+	table := bucket.(*readWriteBucket).table
+
+	_, err := tx.tx.Exec(context.TODO(), "DROP TABLE public."+table+";")
```

Comment:
I assume indices will be dropped automatically?

Reply:
Yes they are. Tested to be sure.